### PR TITLE
Add Baoswap xDai liquidity

### PR DESCRIPTION
**Pull request by nlordell**
Follow up on #1403 

This PR adds Baoswap liquidity for xDai. Not much (since xDai just doesn't have that much liquidity) but may help give some better prices.

I added a macro to implement new Uniswap-like liquidity to cut down on boilerplate.

### Test Plan

Added a unit test to verify contract address for [wxDAI](https://blockscout.com/poa/xdai/tokens/0xe91d153e0b41518a2ce8dd3d7944fa863463a97d)-[1INCH](https://blockscout.com/poa/xdai/tokens/0x7f7440c5098462f833e123b44b8a03e1d9785bab) pair: [`0x8746355882E10AAE144d3709889dfAA39FF2a692`](https://blockscout.com/xdai/mainnet/address/0x8746355882E10AAE144d3709889dfAA39FF2a692).

Additionally some manual testing:

<details><summary>Run an xDai solver with Baoswap liquidity</summary>

```
$ cargo run -p solver -- \
  --baseline-sources Baoswap \
  --orderbook-url https://protocol-xdai.gnosis.io \
  --node-url "https://rpc.xdaichain.com" \
  --settle-interval 30 \
  --solver-account 0x97d69672E8FE5be64D7D5BbbA438ae9b08187667 \
  --transaction-strategy DryRun
...
2021-11-18T12:35:23.080Z  INFO solver::settlement_submission::dry_run: not submitting transaction in dry-run mode
2021-11-18T12:35:23.080Z DEBUG solver::settlement_submission::dry_run: transaction simulation: https://dashboard.tenderly.co/gp-v2/staging/simulator/new?block=19132854&blockIndex=0&from=0x97d69672e8fe5be64d7d5bbba438ae9b08187667&gas=8000000&gasPrice=0&value=0&contractAddress=0x9008d19f58aabd9ed0d60971565aa8510560ab41&network=100&rawFunctionInput=0x13d79a0b000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000014000000000000000000000000000000000000000000000000000000000000007600000000000000000000000000000000000000000000000000000000000000002000000000000000000000000ddafbb505ad214d7b80b1f830fccc89b60fb7a83000000000000000000000000e91d153e0b41518a2ce8dd3d7944fa863463a97d00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000dd90b77590026a900000000000000000000000000000000000000000000000000000000000f360000000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000004200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000022e54af87160f6dd1400ff28a3af2cb157d17931000000000000000000000000000000000000000000000000000000003b358305000000000000000000000000000000000000000000000035de80db5c4906f34900000000000000000000000000000000000000000000000000000000619648ef0000000000000000000000000000000000000000000000000000000000000ccc00000000000000000000000000000000000000000000000000000000000008740000000000000000000000000000000000000000000000000000000000000021000000000000000000000000000000000000000000000035de80db5c4906f3490000000000000000000000000000000000000000000000000000000000000160000000000000000000000000000000000000000000000000000000000000004167bc3cbff769b5dc9f5658a7db4f58ea83aea9ce8713d25b8ddd9d392a904f2950a3bee7f63e23538c7f424bbe80276ac79532a5d1d17260083f9f7f50adcdd31b00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000003e6a0d5b96e587496c54cddcb068eee5ad52b21a000000000000000000000000000000000000000000000035de80db5c4906f349000000000000000000000000000000000000000000000000000000003ab0e5eb0000000000000000000000000000000000000000000000000000000061964f66487b02c558d729abaf3ecf17881a4181e5bc2446429a0995142297e897b6eb370000000000000000000000000000000000000000000000000007ab3c4e63d9570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000035de80db5c4906f349000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000000413f28ee4d0451a8784d35636be7e1bd1897261ba53f447fcb0d3ad06c66561cdf7254231f54ea07ffee664a37671247d612eb5a411462bdfa4be838a6c3710e771c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000fdde383a132d858408b7f2460fea5ca6c571d0420000000000000000000000000000000000000000000000000dd90b77590026a900000000000000000000000000000000000000000000000000000000000dcae30000000000000000000000000000000000000000000000000000000061964f73487b02c558d729abaf3ecf17881a4181e5bc2446429a0995142297e897b6eb370000000000000000000000000000000000000000000000000007ab3c4e63d95700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000dd90b77590026a9000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000000412a9cdc2c87aaefaabddec3939793c4ef2edb3a2008a42b9d66213f76bb1684797cf44c8accba1494e33c7d3ea82ba0744b39fbe006e9b8837474ab706e89d7061b000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000003600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000120000000000000000000000000e91d153e0b41518a2ce8dd3d7944fa863463a97d000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000044095ea7b30000000000000000000000006093aebac87d62b1a5a4ceec91204e35020e38beffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000000000000000000000000000000000000000000000000000000000006093aebac87d62b1a5a4ceec91204e35020e38be0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000001048803dbee00000000000000000000000000000000000000000000000000000000000f36010000000000000000000000000000000000000000000000000ddc96ff672b6a4900000000000000000000000000000000000000000000000000000000000000a00000000000000000000000009008d19f58aabd9ed0d60971565aa8510560ab41ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000e91d153e0b41518a2ce8dd3d7944fa863463a97d000000000000000000000000ddafbb505ad214d7b80b1f830fccc89b60fb7a83000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
2021-11-18T12:35:23.080Z  INFO solver::driver: Successfully submitted NaiveSolver settlement: 0x0000000000000000000000000000000000000000000000000000000000000000
...
```

</details>

Check the [simulation link](https://dashboard.tenderly.co/gp-v2/staging/simulator/11793496-b526-4afa-a044-4eb7a2ba78d6) swaps with the Baoswap router [`0x6093AeBAC87d62b1A5a4cEec91204e35020E38bE`](https://blockscout.com/xdai/mainnet/address/0x6093AeBAC87d62b1A5a4cEec91204e35020E38bE).

### CODE REVIEW

File: crates/contracts/build.rs [hunk 1]
```diff
@@ -149,36 +155,36 @@ fn main() {
...
-            .add_network_str("4", "0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506")
-            .add_network_str("100", "0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506")
-    });
     generate_contract_with_config("SushiSwapFactory", |builder| {
         builder
             .add_network_str("1", "0xC0AEe478e3658e2610c5F7A4A2E1777cE9e4f2Ac")
             .add_network_str("4", "0xc35DADB65012eC5796536bD9864eD8773aBc74C4")
             .add_network_str("100", "0xc35DADB65012eC5796536bD9864eD8773aBc74C4")
     });
-    generate_contract_with_config("UniswapV2Router02", |builder| {
```
Comment (bh2smith) [inline crates/contracts/build.rs:172]: I guess this was a rename...

File: crates/contracts/src/lib.rs [hunk 2]
```diff
@@ -3,18 +3,20 @@ pub mod paths;
...
     env!("OUT_DIR"),
-    "/BalancerV2WeightedPoolFactory.rs"
+    "/BalancerV2WeightedPool2TokensFactory.rs"
 ));
 include!(concat!(
     env!("OUT_DIR"),
-    "/BalancerV2WeightedPool2TokensFactory.rs"
+    "/BalancerV2WeightedPoolFactory.rs"
 ));
-include!(concat!(env!("OUT_DIR"), "/BalancerV2StablePoolFactory.rs"));
```
Comment (bh2smith) [inline crates/contracts/src/lib.rs:17]: Were these just reordered for fun then? Doesn't appear to have been alphabetized...

**Review by bh2smith — APPROVED** [review]
Awesome! No more terrible swaps...

File: crates/contracts/build.rs [hunk 3]
```diff
@@ -115,6 +115,12 @@ fn main() {
             "on_swap_with_balances"
         )
     });
+    generate_contract_with_config("BaoswapFactory", |builder| {
+        builder.add_network_str("100", "0x45DE240fbE2077dd3e711299538A09854FAE9c9b")
+    });
+    generate_contract_with_config("BaoswapRouter", |builder| {
+        builder.add_network_str("100", "0x6093AeBAC87d62b1A5a4cEec91204e35020E38bE")
+    });
```
Comment (nlordell) [inline crates/contracts/build.rs:123]: The only real changes in this file, the other are just `sort`-ing by contract name.

File: crates/contracts/src/lib.rs (hunk 2 above)
Comment (nlordell) [reply crates/contracts/src/lib.rs:17]: The are in alphabetical order now, they weren't before.

**Review by vkgnosis — APPROVED** [review]

**Review by fleupold — COMMENTED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia3.giphy.com%2Fmedia%2FWtj1qC5B0pO5G%2Fgiphy-downsized-medium.gif&token=amazeballs" data-gitmeme-token="amazeballs"></a>

Next up, SwapR (Uniswap clone from dxDAO)

**Comment by nlordell** [comment]
Ask and thou shalt receive: https://github.com/gnosis/gp-v2-services/pull/1409

**Pull request merged**